### PR TITLE
Add JSON schema generation

### PR DESCRIPTION
**Pull request by armintaenzertng**
This adds some basic code to generate a json schema for the model.

### CODE REVIEW

**Comment by zvr** [comment]
I don't think spec-parser should have this functionality.

If we can generate an RDF ontology, we already have the tools to automatically translate it and generate all the other schemas, as we have been doing it for years in SPDXv2.

**Comment by armintaenzertng** [comment]
I'd be very interested in a tool that can parse the current ontology and convert it into a json schema!

The only thing I've come across so far is [Gary's code in the java-tools for SPDX2](https://github.com/spdx/tools-java/blob/master/src/main/java/org/spdx/tools/schema/OwlToJsonSchema.java), which is specialized on dealing with SPDX2 properties (see for example [here](https://github.com/spdx/tools-java/blob/974aad5bfe4afb222e8f01044e88e8252a60b60d/src/main/java/org/spdx/tools/schema/OwlToJsonSchema.java#L112)). So I figured that instead of parsing in the information that we just translated into the OWL/SHACL, it would be best to start at the source of the information and directly generate a json schema from there.

As I said, if you know of a tool that already works with the current OWL, please give me a pointer, that would be much appreciated!

**Comment by goneall** [comment]
I haven't had time to look at the PR yet - but I was also thinking of generating the JSON schema directly from the markdown source.  We can also update the Java tools to work with the OWL document - it should only require minor changes.

I'll take a look at the PR in the next day or two and refresh my memory on the OWL to JSON Schema code and update this PR with some additional thoughts.

File: spec_parser/utils.py [hunk 1]
```diff
@@ -162,6 +163,123 @@ def gen_rdf(self) -> None:
...
+
+    def _spdx_type_to_json_type(self, type_name: str) -> str:
+        if type_name in ["xsd:string", "xsd:anyURI"]:
+            return "string"
+        if type_name in ["xsd:nonNegativeInteger", "xsd:positiveInteger", "xsd:decimal"]:
+            return "number"
+        if type_name in ["xsd:boolean"]:
+            return "bool"
+
+        print("No conversion to json type implemented for " + type_name)
```
Comment (goneall) [inline spec_parser/utils.py:281]: We should probably do more than just print an error - perhaps set an error code for the CI or throw an exception.

File: spec_parser/utils.py [hunk 2]
```diff
@@ -162,6 +163,123 @@ def gen_rdf(self) -> None:
         with safe_open(fname, "w") as f:
             f.write(g.serialize(format="turtle"))
 
+    def gen_schema(self) -> None:
```
Comment (goneall) [inline spec_parser/utils.py:166]: Do we include the description of the classes and fields in the schema?  If not, we may want to add those.

File: spec_parser/utils.py [hunk 3]
```diff
@@ -162,6 +163,123 @@ def gen_rdf(self) -> None:
...
+        schema = {
+            "$schema": "https://spdx.org/spdx3/json_schema",  # TODO: insert correct schema and id URIs
```
Comment (goneall) [inline spec_parser/utils.py:168]: You should use the same version as the current schema - there was feedback/issues on using the most relevant version.

**Review by goneall — CHANGES_REQUESTED** [review]
A few initial suggestions - I'll need to do a more detailed review and review the output before having a complete set of suggestions.

**Comment by goneall** [comment]
After reviewing the Java code and the PR, I'm leaning towards adding a JSON schema generator using the input markdown.  There are two reasons (in priority order):
- The Java code does not interpret the SHACL code - adding that would be a small to moderate amount of work, but likely less work than completing the PR provided here
- The markdown code is closer to the original data, so it is less likely that errors would be introduced by the intermediate translations

File: spec_parser/utils.py (hunk 2 above)
Comment (armintaenzertng) [reply spec_parser/utils.py:166]: The output is already 5000 lines long, so I decided to omit descriptions. If there are use-cases where they are actually needed, I can add them in, though.

File: spec_parser/utils.py (hunk 1 above)
Comment (armintaenzertng) [reply spec_parser/utils.py:281]: Currently, `context` is the (only) thing that trips this. I agree that this should raise an error, I will add that once the `context` is dealt with.

Comment (armintaenzertng) [reply spec_parser/utils.py:281]: Sorry, I meant `extension` instead of `context`.

**Comment by maxhbr** [comment]
Due to the refactoring this is no longer can be merged. We might want to close it.

**Comment by armintaenzertng** [comment]
Closing due to outdatedness.

**Pull request closed without merging**